### PR TITLE
WIP: create dao info page alpha

### PR DESCRIPTION
**Pull request by scrappydew**
page needs to have the event max listener issue fixed, then this can be merged for alpha release

### CODE REVIEW

**Comment by tentacle-fi** [comment]
Suggested improvements:

1. Skeleton loading elements for the DAO page
2. Tooltips for each item to explain what it's value means. i'm currently confused about why the page shows 0 LP standing

File: src/utils/index.ts [hunk 1]
```diff
@@ -248,6 +264,150 @@ export interface ICurrentStats {
...
+}
+
+export const getDaoHoldings = async (provider: provider): Promise<IDaoHoldings> => {
+  // Get coin and token holdings from the DAO multisig address and farming address
+  const ubqHoldings = (await getCoinBalanceAsBigNum(provider, DAO_MULTISIG)).plus(await getCoinBalanceAsBigNum(provider, DAO_FARMING));
+  const inkHoldings = (await getBalanceAsBigNum(provider, INK, DAO_MULTISIG)).plus(await getBalanceAsBigNum(provider, INK, DAO_FARMING));
+  const eschHoldings = (await getBalanceAsBigNum(provider, ESCH, DAO_MULTISIG)).plus(await getBalanceAsBigNum(provider, ESCH, DAO_FARMING));
+
+  // Get lp holdings from the DAO multisig address and farming address
+  const ubqInkHoldings = (await getBalanceAsBigNum(provider, INK_UBQ_LPAddress, DAO_MULTISIG)).plus(
```
Comment (tentacle-fi) [inline src/utils/index.ts:293]: ideally for the future we'd make this so it's an array of the values we want to acquire. and do a 'foreeach' of the elements in the array, so the 'get' code doesn't have to be modified but the array does.

File: src/utils/index.ts [hunk 2]
```diff
@@ -248,6 +264,150 @@ export interface ICurrentStats {
...
+      ubqInk: ubqInkHoldings,
+      gransInk: gransInkHoldings,
+      inkEsch: inkEschHoldings,
+    },
+  } as IDaoHoldings;
+};
+
+// Calculates the circulating INK supply based on the amount still held by
+// the DAO minting address plus a sum of all of the individual farm holdings
+export const getCirculatingSupply = async (provider: provider): Promise<ICirculatingSupply> => {
```
Comment (tentacle-fi) [inline src/utils/index.ts:317]: is there a check somewhere to ensure, at runtime, that the provider is set?

File: src/utils/index.ts [hunk 3]
```diff
@@ -248,6 +264,150 @@ export interface ICurrentStats {
...
+
+  // Grab all the individual INK quantities
+  let heldByDeployer = await getBalanceAsBigNum(provider, INK, INKMINTEDHOLDINGADDRESS);
+  let heldByUBQINK = await getBalanceAsBigNum(provider, INK, INK_UBQ_FarmContract);
+  let heldByGRANSUBQ = await getBalanceAsBigNum(provider, INK, INK_GRANS_FarmContract);
+  let heldByUBQESCH = await getBalanceAsBigNum(provider, INK, INK_ESCH_FarmContract);
+  const subtotal = heldByDeployer.plus(heldByUBQINK).plus(heldByGRANSUBQ).plus(heldByUBQESCH).toString();
+
+  // Total them up, and subtract them from the totalSupply
+  let circulatingTotal = totalSupply.minus(subtotal);
```
Comment (tentacle-fi) [inline src/utils/index.ts]: const not let

File: src/utils/index.ts [hunk 4]
```diff
@@ -248,6 +264,150 @@ export interface ICurrentStats {
...
+  const currentBlock = await web3.eth.getBlockNumber();
+
+  try {
+    // TODO: Perhaps we should just store these in available farms?
+    const INKCONTRACT = getERC20Contract(provider, INK);
+    const UBQINKFARMCONTRACT = getFarmContract(provider, INK_UBQ_FarmContract);
+    const GRANSINKFARMCONTRACT = getFarmContract(provider, INK_GRANS_FarmContract);
+    const INKESCHFARMCONTRACT = getFarmContract(provider, INK_ESCH_FarmContract);
+
+    // TODO: any way to make this more flexible using availablefarms?
```
Comment (tentacle-fi) [inline src/utils/index.ts:383]: this can easily be an array that uses foreach, like the other suggestion.

File: src/utils/index.ts [hunk 5]
```diff
@@ -122,6 +132,12 @@ export const getERC20Contract = (provider: provider, address: string) => {
   return contract;
 };
 
+export const getFarmContract = (provider: provider, address: string) => {
```
Comment (tentacle-fi) [inline src/utils/index.ts:135]: this does not need to exist. it's a copy and farm contracts are the same as erc20 contracts, from what i can tell. if we need to extend the 'get contract' function we can send in the ABI info to ensure it's more flexible.

if there is a good reason i'm wrong, please state that.

File: src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// import Tooltip from "@mui/material/Tooltip";
+// import useUbiq from "hooks/useUbiq";
+import { getDaoHoldings, IDaoHoldings } from "utils";
+import { useWallet } from "use-wallet";
+import useUbiq from "hooks/useUbiq";
+import { provider } from "web3-core";
+import BigNumber from "bignumber.js";
+
+// sets up a formatter so with toFormat on big numbers, we get thousands separators
+BigNumber.config({ FORMAT: { groupSeparator: ",", groupSize: 3, decimalSeparator: "." } });
```
Comment (tentacle-fi) [inline src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx:18]: if we're going to use this format a LOT, we should consider exporting this object from utils or another common place.

File: src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      const currentHoldings = await getDaoHoldings(provider);
+
+      setDaoHoldings(currentHoldings);
+    },
+    [setDaoHoldings]
+  );
+
+  // Update holdings based on new block numbers
+  useEffect(() => {
+    fetchDaoHoldings(ethereum);
```
Comment (tentacle-fi) [inline src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx:39]: typically the pattern for 'ethereum' is to check if it exists before calling the function eg

`if(!ethereum){ return }`

File: src/views/DaoInfo/components/StatsRibbon/StatsRibbon.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  }, [ethereum, setCirculatingSupply]);
+
+  const fetchCurrentMarketcap = useCallback(async () => {
+    if (tokenPrices === undefined) {
+      return;
+    }
+
+    if (circulatingSupply === undefined) {
+      return;
+    }
```
Comment (tentacle-fi) [inline src/views/DaoInfo/components/StatsRibbon/StatsRibbon.tsx]: combine these two into one single if statement with OR logic

**Review by tentacle-fi — DISMISSED** [review]
see comments for changes. looks pretty good. we can look at using or extending available farms or making a new source for the dao's tokens to make the logic array based.

File: src/utils/index.ts (hunk 4 above)
Comment (scrappydew) [reply src/utils/index.ts:383]: fixed

File: src/views/DaoInfo/components/StatsRibbon/StatsRibbon.tsx (hunk 8 above)
Comment (scrappydew) [reply src/views/DaoInfo/components/StatsRibbon/StatsRibbon.tsx]: fixed

File: src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx (hunk 7 above)
Comment (scrappydew) [reply src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx:39]: fixed

**Comment by scrappydew** [comment]
> 
>     2. Tooltips for each item to explain what it's value means. i'm currently confused about why the page shows 0 LP standing

i intend on converting the info icon im already using on the farm page for this, but its acting strange. ill show you when i review a couple of these other items with you

File: src/utils/index.ts (hunk 2 above)
Comment (scrappydew) [reply src/utils/index.ts:317]: it doesn't look like they've ever done that in this file, so i don't know how that works

File: src/utils/index.ts (hunk 4 above)
Comment (tentacle-fi) [reply src/utils/index.ts:383]: issue created

File: src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx (hunk 6 above)
Comment (tentacle-fi) [reply src/views/DaoInfo/components/DAOHoldingsDetails/DAOHoldingsDetails.tsx:18]: issue created

**Pull request merged**